src/nginx.c

```diff
@@ -73,108 +85,261 @@ static size_t nginx_curl_callback(void *buf, size_t size, size_t nmemb,
+
+  if ((ngx_inst[instance_num_counter].curl = curl_easy_init()) == NULL) {
+    ERROR("nginx plugin: curl_easy_init failed.");
+    return -1;
+  }
+
+  return 0;
+}
+/******************/
+
+/******* Nginx plugin config parsing. Inspired by ceph plugin config parser
+ * implementation ( see ceph.c ) *******/
+static int nginx_handle_str(struct oconfig_item_s *item, char *dest,
+                            int dest_len) {
+  const char *val;
+  if (item->values_num != 1) {
+    return -ENOTSUP;
+  }
+  if (item->values[0].type != OCONFIG_TYPE_STRING) {
+    return -ENOTSUP;
+  }
+  val = item->values[0].value.string;
+  if (snprintf(dest, dest_len, "%s", val) > (dest_len - 1)) {
+    ERROR("nginx plugin: configuration parameter '%s' is too long.\n",
+          item->key);
+    return -ENAMETOOLONG;
+  }
   return 0;
-} /* void init */
+}
+
+static int nginx_add_daemon_config(oconfig_item_t *ci) {
+  int ret;
+
+  /* Get instance name first - we need it for building metrics prefix as well */
+  if ((ci->values_num != 1) || (ci->values[0].type != OCONFIG_TYPE_STRING)) {
+    WARNING("nginx plugin: `InstanceName' block must have only name as an "
+            "argument.");
+    return -1;
+  }
 
-static void submit(const char *type, const char *inst, long long value) {
+  ret = nginx_handle_str(ci, ngx_inst[instance_num_counter].name,
+                         DATA_MAX_NAME_LEN);
+  if (ret) {
+    return ret;
+  }
+
+  if (ngx_inst[instance_num_counter].name[0] == '\0') {
+    ERROR("nginx plugin: you must configure nginx instance name.\n");
+    return -EINVAL;
+  }
+  /*****************************************************************************/
+
+  ret = init_curl_structure();
+  if (ret) {
+    return ret;
+  }
+  set_default_curl_opts(); // Set default curl params for nginx instance
+
+  init_nginx_instance_settings_array(); // Initialize stringlist berore parsing
+                                        // collectd config for nginx instance
+
+  for (int i = 0; i < ci->children_num; i++) {
+    oconfig_item_t *child = ci->children + i;
+
+    for (int j = 0; j < config_keys_num; j++) {
+      if (strcasecmp(config_keys[j], child->key) == 0) {
+        ret = nginx_handle_str(child, parsed_values_for_nginx_instance[j],
+                               DATA_MAX_NAME_LEN);
+        if (ret) {
+          return ret;
+        }
+        break;
+      }
+    }
+  }
+
+  ret =
+      set_optional_curl_params(); // Set optional curl params for nginx instance
+  if (ret) {
+    return ret;
+  }
+
+  instance_num_counter++; // Increase counter of configured nginx instances.
+
+  return 0;
+}
+
+static int nginx_config(oconfig_item_t *ci) {
+  int ret;
+
+  parsed_values_for_nginx_instance =
+      (char **)calloc(config_keys_num, sizeof(char *));
+  for (int i = 0; i < config_keys_num; i++) {
+    parsed_values_for_nginx_instance[i] =
+        (char *)calloc(DATA_MAX_NAME_LEN, sizeof(char));
+  }
```

Comment:
You have deinit function for freeing the allocs, so there should be also corresponding init function doing the allocs (close together).